app/services/AuthorisationService.scala

```diff
@@ -17,86 +17,54 @@
-import uk.gov.hmrc.auth.core.retrieve.Retrievals._
 import uk.gov.hmrc.auth.core.retrieve._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals._
 import uk.gov.hmrc.http.HeaderCarrier
 
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
 class AuthorisationService @Inject()(
-  val authConnector: AuthConnector,
-  val userDetailsConnector: UserDetailsConnector,
-  val taxEnrolmentService: TaxEnrolmentService) extends AuthorisedFunctions {
+                                      val authConnector: AuthConnector,
+                                      val taxEnrolmentService: TaxEnrolmentService) extends AuthorisedFunctions {
 
-  def enrolmentAuthorised(implicit hc: HeaderCarrier): Future[Either[Boolean, String]] = {
-    authorised(Enrolment("HMRC-LISA-ORG")).retrieve(authorisedEnrolments) { enr =>
-      enr.getEnrolment("HMRC-LISA-ORG") match {
-        case None => {
-          Logger.warn("Authorised but no enrolment object")
-          Future.successful(Left(false))
-        }
-        case Some(e) => {
-          Logger.info("Got enrolment object for HMRC-LISA-ORG")
-          Future.successful(Right(e.getIdentifier("ZREF").get.value))
-        }
-      }
-    } recoverWith {
-      case _ => {
-        Logger.warn("The enrolment is not authorised")
-        Future.successful(Left(false))
-      }
-    }
-  }
-
-  def userStatus(implicit hc: HeaderCarrier): Future[LisaUserStatus] = {
+  def userStatus(implicit hc: HeaderCarrier): Future[UserStatus] = {
     authorised(
       AffinityGroup.Organisation and AuthProviders(GovernmentGateway)
-    ).retrieve(internalId and userDetailsUri) { case (id ~ uri) =>
-      val userId = id.getOrElse(throw new RuntimeException("No internalId for user"))
-      val userUri = uri.getOrElse(throw new RuntimeException("No userDetailsUri for user"))
-
-      userDetailsConnector.getUserDetails(userUri)(hc) flatMap { user =>
-        val groupId = user.groupIdentifier.getOrElse(throw new RuntimeException("No groupIdentifier for user"))
-
-        enrolmentAuthorised(hc) flatMap { res =>
-          res match {
-            case Right(zref) => {
-              Logger.info("HMRC-LISA-ORG Enrolment is Authorised")
-              Future.successful(UserAuthorisedAndEnrolled(userId, user, zref))
-            }
-            case Left(_) => {
-
-              Logger.info("The enrolment has not been authorised yet so checking Enrolments service the group id is " + groupId)
-              taxEnrolmentService.getNewestLisaSubscription(groupId)(hc) map {
+    ).retrieve(internalId and groupIdentifier and authorisedEnrolments) {
+      case Some(id) ~ Some(groupId) ~ enrolments =>
+        statusFromAuth(id, enrolments)
+          .map(a => Future.successful(Some(a)))
+          .getOrElse(statusFromTaxEnrolments(id, groupId))
+          .map(_.getOrElse(UserAuthorised(id, TaxEnrolmentDoesNotExist)))
+      case _ => Future.successful(UserUnauthorised)
+    } recover {
```

Comment:
Curious, what does recover do?

Reply:
Easiest to think of it as a bit like a `catch` in other languages - it's a partial function for handling Throwables.